keps/sig-auth/3331-structured-config-for-oidc-authentication/README.md

```diff
@@ -0,0 +1,638 @@
+    
+    - rule: 'claims.roles.exists(r, r == "kubernetes-user")'
+      message: only kubernetes-user group members can access the cluster
+    ```
+
+3. `ClaimMappings` - rules to map claims from a token to Kubernetes user attributes.
+    ```go
+    type UserAttributes struct {
+        // Username represents an option for the username attribute.
+        Username string
+        // Groups represents an option for the groups attribute. 
+        // +optional
+        Groups string
+        // UID represents an option for the uid attribute.
+        // +optional
+        UID string
+        // Extra represents an option for the extra attribute.
+        // +optional
+        Extra []ExtraMapping
+    }
+   
+    type ExtraMapping struct {
+        // Key is a CEL expression to extract extra attribute key.
+        Key string
+        // Expression is a CEL expression to extract extra attribute value.
+        Expression string
+    }
+    ```
+
+    Every field of structures above is a CEL expression. For username and uid, the expression should evaluate a string.
+    For groups, it should evaluate to an array. A map of extra attributes is composed of key/value pairs, each with their
+    own CEL expressions, both should evaluate to a string.
+
+    The example of mapping user attributes can be found below:
+
+    ```yaml
+    claimMappings:
+      username: 'claims.username + ":external-user"'
+      groups: 'claims.roles.split(",")'
+      uid: 'claims.sub'
+      extra:
+      - key: '"client_name"'
+        expression: 'claims.aud'
+    ```
+   
+    For the token with the following claims
+    ```json
+    {
+      "sub": "119abc",
+      "aud": "kubernetes",
+      "username": "jane_doe",
+      "roles": "admin,user",
+      ...
+    }
+    ```
+    our authenticator will extract the following user attributes:
+    ```yaml
+    username: jane_doe:external-user
+    groups: ["admin", "user"]
+    uid: "119abc"
+    extra:
+      client_name: kubernetes
+    ```
+4. `ClaimsFilter` - list of claim names that should be passed to CEL expressions. The assumption is that administrators
+   know the structure of the token and the exact claims they will use in CEL expressions.
+   This option helps to reduce system load and operate only with required claims.
+
+### More about CEL
+
+* CEL runtime should be compiled only once if structured OIDC config option is enabled
+* To make working with strings more convinient, `strings` and `encoding` [CEL extensions](https://github.com/google/cel-go/tree/v0.9.0/ext) should be enabled, 
+e.g, to be able to split a string with comma separated fields and use them as a single array
+* Benchmarks are required to see how different CEL expressions affects authentication time
+* CEL expressions are called on each request. We should properly investigate the influence of these calls on the system
+  latency and, if necessary, prove caching or other mechanisms to improve performance.
```

Comment:
But we would still have the 10 second token cache on the tokens.  Also, I think we can do better here by making the token cache last until the expiration of the token or if the public keys change, whichever happens first.